src/RandMsg.chpl

```diff
@@ -572,70 +573,117 @@ module RandMsg
-                const generatorIdxOffset = here.id * nTasksPerLoc,
-                    locSubDom = poissonArr.localSubdomain(),  // the chunk that this locale needs to handle
-                    indicesPerTask = locSubDom.size / nTasksPerLoc;  // the number of elements each task needs to handle
-
-                coforall tid in Tasks {
-                    const taskSeed = generatorSeed + generatorIdxOffset + tid,  // initial seed offset by other locales threads plus current thread id
-                        startIdx = tid * indicesPerTask,
-                        stopIdx = if tid == nTasksPerLoc - 1 then locSubDom.size else (tid + 1) * indicesPerTask;  // the last task picks up the remainder of indices
-                    var rs = new randomStream(real, taskSeed);
-                    for i in startIdx..<stopIdx {
-                        var L = exp(-lam);
-                        var k = 0;
-                        var p = 1.0;
-
-                        do {
-                            k += 1;
-                            p = p * rs.next(0, 1);
-                        } while p > L;
-                        poissonArr[locSubDom.low + i] = k - 1;
+        if hasSeed {
+            // use a fixed number of elements per stream instead of relying on number of locales or numTasksPerLoc because these
+            // can vary from run to run / machine to mahchine. And it's important for the same seed to give the same results
+            const generatorSeed = (rng.next() * 2**62):int,
+                elemsPerStream = 4096;
+
+            if isSingleLam {
+                const lam = lamStr:real;
+                // using nested coforalls over locales and tasks so we know how to generate taskSeed
+                coforall loc in Locales do on loc {
+                    const locSubDom = poissonArr.localSubdomain(),
+                        offset = if loc.id == 0 then 0 else elemsPerStream - (locSubDom.low % elemsPerStream);
+
+                    // skip if all the values were pulled to previous locale
+                    if offset <= locSubDom.high {
+                        // we take the ceil in chunk calculation because if elemsPerStream doesn't evenly divide along locale boundaries, the remainder is pulled to the previous locale
+                        const chunksAlreadyDone = if loc.id == 0 then 0 else ceil((locSubDom.low + 1) / elemsPerStream:real):int,  // number of chunks handled by previous locales
+                            thisLocsNumChunks = ceil((locSubDom.high + 1 - locSubDom.low + offset) / elemsPerStream:real):int;  // number of chunks this locale needs to handle
+
+                        coforall streamID in 0..<thisLocsNumChunks {
+                            const taskSeed = generatorSeed + chunksAlreadyDone + streamID,  // initial seed offset by other locales threads plus current thread id
+                                startIdx = (streamID * elemsPerStream) + offset,
```

Comment:
Could you move `+ locSubDom.low` up to this line? (I think it might make it more clear that `offset` is a separate concept that's not accounting for where you are starting in the global array)